### PR TITLE
Add theory of (infinite) sets

### PR DESCRIPTION
**Pull request by xldenis**
Adds a theory of (infinite) sets.

We should also add `FSet` the theory of finite sites which notably would have an induction principle and a cardinality.

### CODE REVIEW

**Comment by jhjourdan** [comment]
Would it make sense to have a type of both finite and infinite sets, and a function of cardinality which returns an option (i.e., None when the set is infinite)?

Then it's easy to do induction on the finite set, by assuming that the cardinality is some integer.

**Comment by jhjourdan** [comment]
(I don't know whether this is a good idea or not, I'm just suggesting.)

File: creusot-contracts/src/logic/set.rs [hunk 1]
```diff
@@ -9,15 +9,54 @@ use std::ops::Index;
...
-    #[creusot::builtins = "set.Set.add"]
+    #[creusot::builtins = "set.Set.mem"]
+    fn mem(e: T, set: Self) -> bool {
+        pearlite! { absurd }
+    }
+
+    #[logic]
+    #[why3::attr = "inline:trivial"]
     pub fn insert(self, e: T) -> Self {
+        Self::add(e, self)
```
Comment (jhjourdan) [inline creusot-contracts/src/logic/set.rs:30]: I'm not such a great fan of having plenty of notions which are aliases from each other: insert/add remove/rem mem/contains. Why do we need these aliases?

Comment (xldenis) [reply creusot-contracts/src/logic/set.rs:30]: `mem` is private to the `Set` module, it is not exported as part of the API. It's needed as must match the signature of the `why3` function which is flipped wrt Rust conventions.

**Comment by xldenis** [comment]
> Would it make sense to have a type of both finite and infinite sets, and a function of cardinality which returns an option (i.e., None when the set is infinite)?

That's an interesting idea, Why3 does this by adding a second module `Cardinal` which provides an `is_finite` predicate a set can have and then allows access to the cardinality. 

I wonder if there are underlying reasons for this? Perhaps we should just avoid depending on why3 theories entirely.

**Comment by xldenis** [comment]
I'm going to merge this for now, we can revisit it to add finite sets (either via optional cardinality or seperate type).

**Comment by xldenis** [comment]
Problem: this seems to break the red black tree proofs. it might be necessary to be a bit smarter about `use set.Set` in `Types`...

**Pull request merged**